In [1]:
import pandas as pd

## Reading files

In [2]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [3]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [4]:
wh = pd.concat([wh2018,wh2019])

In [5]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [6]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [29]:
x=wh.isnull()
x

Overall rank  Country or region  Score  GDP per capita  Social support  \
0           False              False  False           False           False   
1           False              False  False           False           False   
2           False              False  False           False           False   
3           False              False  False           False            True   
4           False              False  False           False           False   
..            ...                ...    ...             ...             ...   
151         False              False  False           False           False   
152         False              False  False           False           False   
153         False              False  False           False           False   
154         False              False  False           False           False   
155         False              False  False           False           False   

     Healthy life expectancy  Freedom to make life choices  Generosity  \
0                      False                         False       False   
1                      False                         False       False   
2                      False                         False       False   
3                      False                         False        True   
4                      False                         False       False   
..                       ...                           ...         ...   
151                    False                         False       False   
152                    False                         False       False   
153                    False                         False       False   
154                    False                         False       False   
155                    False                         False       False   

     Perceptions of corruption   year  
0                        False  False  
1                        False  False  
2                        False  False  
3                        False  False  
4                        False  False  
..                         ...    ...  
151                      False  False  
152                      False  False  
153                      False  False  
154                      False  False  
155                      False  False  

[312 rows x 10 columns]

In [30]:
prom = x.groupby(['year']).mean()
print(prom)

       Overall rank  Country or region     Score  GDP per capita  \
year                                                               
False           0.0                0.0  0.060897        0.038462   

       Social support  Healthy life expectancy  Freedom to make life choices  \
year                                                                           
False        0.051282                 0.083333                      0.073718   

       Generosity  Perceptions of corruption  
year                                          
False    0.067308                   0.041667  


In [38]:
xx=wh.groupby(['year']).mean()
xx

Overall rank     Score  GDP per capita  Social support  \
year                                                           
2018          78.5  5.354060        0.892748        1.205915   
2019          78.5  5.434028        0.900758        1.208265   

      Healthy life expectancy  Freedom to make life choices  Generosity  \
year                                                                      
2018                 0.597576                      0.456514    0.181360   
2019                 0.722965                      0.395441    0.185092   

      Perceptions of corruption  
year                             
2018                   0.112859  
2019                   0.109053

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [36]:
l=wh.groupby(['year']).describe()
l

Overall rank                                                    Score  \
            count  mean        std  min    25%   50%     75%    max  count   
year                                                                         
2018        156.0  78.5  45.177428  1.0  39.75  78.5  117.25  156.0  151.0   
2019        156.0  78.5  45.177428  1.0  39.75  78.5  117.25  156.0  142.0   

                ... Generosity        Perceptions of corruption            \
          mean  ...        75%    max                     count      mean   
year            ...                                                         
2018  5.354060  ...    0.23400  0.598                     149.0  0.112859   
2019  5.434028  ...    0.24825  0.566                     150.0  0.109053   

                                                     
           std  min      25%    50%      75%    max  
year                                                 
2018  0.098108  0.0  0.05000  0.082  0.13800  0.457  
2019  0.093279  0.0  0.04775  0.085  0.14025  0.453  

[2 rows x 64 columns]

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

In [42]:
df=wh['year'].mean()
df


2018.5

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [7]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro